### ISSUE TITLE
-export-symbols $(srcdir)/link-grammar.def doesn't work?!

### ISSUE THREAD

**Issue by ampli** [opencog/link-grammar#1169]
It seems it is translated just fine by `libtool` into the appropriate `ld` directive.
Using `make V=1`:
`libtool: link: g++  ...  -Wl,-retain-symbols-file -Wl,/usr/local/src/link-grammar-devel/generate/link-grammar/link-grammar.def -o .libs/liblink-grammar.so.5.8.1`

However:
```
$ nm -D link-grammar/.libs/liblink-grammar.so
                 U abort
0000000000023e60 T add_empty_word
000000000000fc30 T afclass_init
0000000000070be0 D afdict_classname
000000000000fc70 T afdict_find
000000000000ff10 T afdict_init
000000000000fd10 T affix_list_add
                 U aligned_alloc
...
```
As you can see, all the symbols are exported.
Do you have any idea why?

In any case, instead of `link-grammar.def` I would like to implement the "visibility" attribute of GCC>4.0, as outlined in https://www.gnu.org/software/gnulib/manual/html_node/Exported-Symbols-of-Shared-Libraries.html.
It involves defining `link_public_api` to be `__attribute__((__visibility__("default")))`.

**Comment by linas** [comment]
I know that on Apple, things don't link unless `link-grammar.def` is exactly just right.

I know that on Linux, `link-grammar.def` is not needed; you can throw it away, and all of the grunge surrounding it, and everything will work fine.  The only other OS I know that needed something like this was AIX, a few decades ago. No clue about *BSD or similar.

**Comment by ampli** [comment]
Exporting only the API functions is supposed to have these benefits:
- A reduced chance for symbol clash.
- Reduced shared binary size. In the case of `liblink-grammar.so.5.8.1` on my computer, it is 2464480 vs. 495520 bytes (a 5 times reduction).
- Runtime speedup due to local linking of hidden symbols (no need to use PLT, which allows overriding e.g. by `LG_PRELOAD`).

However, when I actually test it (a 2 lines change), `link-parser` runs 2 times slower (for any corpus), and I have no idea why!
(Of course, I verified that after the said change, "nm -D" shows only the API functions.)
So I am closing this issue and will reopen it if I can make LG faster due to exporting only the API functions.

BTW, there are a few API functions not in `link-grammar.def`, and a few ones in `link-grammar.def` that are not marked with `link_public_api` (thus inaccessible from Windows). I will send a PR to fix that.

**Comment by linas** [comment]
Are you saying that using `__attribute__((__visibility__("default"))).` slowed it down by a factor of two!?

Or are you saying that getting rid of `link-grammar.def` makes it run 2x slower?

If the latter, this is .. surprising. But perhaps explainable as a side-effect of some hidden LTO step. LTO==link-time-optimization -- last time I looked, one had to explicitly turn it on in the linker with `--lto` or -w,lto` or something similar. It would then attempt to in-line any short functions during link-time  (i.e. inline short functions found in one `foo.o` into another `bar.o` and into any executable.)  Perhaps LTO is explicitly turned on if one uses an export file!? Surprising but .. ?

**Comment by linas** [comment]
This:

> a few ones in link-grammar.def that are not marked with `link_public_api` 

There are many such, and they should not be made public. They were placed there because the SAT parser needed them, and it wouldn't link without it.

**Comment by ampli** [comment]
> They were placed there because the SAT parser needed them

There are **no** such functions there by now! The SAT parser is compiled now to a static library (only), in which the LG API calls are **unresolved**.

```
$ nm link-grammar/sat-solver/.libs/libsat-solver.a       
fast-sprintf.o:
0000000000000000 T _Z12fast_sprintfPci
0000000000000060 T _Z12fast_sprintfPcPKc

sat-encoder.o:
                 U __assert_fail
                 U assert_failure
                 U build_disjuncts_for_exp
                 U calloc
                 U check_link_size
                 U clock
                 U compute_link_names
                 U connector_new
...
```

**Comment by ampli** [comment]
> Are you saying that using `__attribute__((__visibility__("default"))).` slowed it down by a factor of two!?

Yes. And the slow down is almost exactly 2!

These are the 2 changes I made (it is really 1-line change and not 2 as I wrote above...):
[nm.zip](https://github.com/opencog/link-grammar/files/6179443/nm.zip)

1.
``` diff
--- a/link-grammar/link-features.h.in
+++ b/link-grammar/link-features.h.in
@@ -24,7 +24,7 @@
 #   define link_public_api(x) __declspec(dllimport) x
 #  endif
 # else
-#  define link_public_api(x) x
+#  define link_public_api(x) __attribute__((visibility("default"))) x
 # endif
 #endif
```
2. Configuring with `CFLAGS=-fvisibility=hidden`.

`nm` then displays only API calls (and unresolved external library calls).
The output is attached.
[nm.zip](https://github.com/opencog/link-grammar/files/6179445/nm.zip)

**Comment by ampli** [comment]
I found the problem:
I forgot that now setting CFLAGS to anything overwrites the optimization (a recent PR contribution).
So I compared it to a run without any optimization (and it's good to know it is only half the speed).

But I try to use `CFLAGS="-O3 -fvisibility=hidden" then POSIX functions are not found!
``` text
.../visibility/link-grammar/utilities.c:64:17: warning: initialization of ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
   64 |  char *errstr = strerror_r(err_no, buf, len);
      |                 ^~~~~~~~~~
.../visibility/link-grammar/utilities.c: In function ‘strtodC’:
.../visibility/link-grammar/utilities.c:763:15: error: implicit declaration of function ‘strtod_l’; did you mean ‘strtodC’? [-Werror=implicit-function-declaration]
  763 |  double val = strtod_l(s, &err, get_C_LC_NUMERIC());
```
It may mean that the `CFLAGS` setting most probably overwrites more settings than it is supposed to (the selection of POSIX functions is a must and is not supposed to be subject to user configuration).

Only tomorrow I will be able to debug that.
(BTW, `strtod_l()` is problematic in any case as I would like to provide my version in one of the next PRs.)

**Comment by linas** [comment]
Try adding `-flto` to `CFLAGS` and see if that changes performance. I'm curious to find out. I expect no more than a percent or two.

**Comment by ampli** [comment]
> Try adding `-flto` to `CFLAGS` and see if that changes performance. I'm curious to find out. I expect no more than a percent or two.

I got about ~0.5% for `en/corpus-basic.batch` and ~1% for  `en/corpus-fix-long.batch`.

**Comment by linas** [comment]
Cool! If you want to add `-flto` permanently, that's OK with me. The only other place where I've used it, it made compiles painfully slow, so I turned it off. For LG, I don't actively compile & develop, so I don't care how slow the compile step is. If it is not annoying to you, you can turn it on.

**Comment by ampli** [comment]
> Cool! If you want to add `-flto` permanently, that's OK with me. The only other place where I've used it, it made compiles painfully slow, so I turned it off. For LG, I don't actively compile & develop, so I don't care how slow the compile step is. If it is not annoying to you, you can turn it on.

A usual compilation (without `-flto`), from scratch, takes on my computer about 2.5 seconds of elapsed time (a fast computer + `ccache`). With `-flto`, it takes 19 seconds (ran twice, to make sure it is cached). The slowness is only due to the linking stage, which takes absolutely most of this time.

The user must be able to override it in case it is buggy or for other reasons, so it should be overridden by user-setting of CFLAGS/CXXFLAGS. It means I can fix the long build-time problem by setting my own CFLAGS. But because package maintainers always use their own flags, they would override it anyway, so the question is one set it in the first place.

To sum up - I can add it (of course with an `autoconf` check that it is supported).
